### PR TITLE
Upgrade compiler tooling and libraries.

### PR DESCRIPTION
**Pull request by vikt0r0**
## Purpose

Upgrade stackage snapshot to `lts-20.4` and ghc to 9.2.5.

## Changes

Change `resolver`in `stack.yaml` and fix build errors and warnings.

## Checklist

- [X] My code follows the style of this project.
- [X] The code compiles without warnings.
- [X] I have performed a self-review of the changes.
- [X] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: haskell-tests/Types/PayloadSerializationSpec.hs [hunk 1]
```diff
@@ -62,7 +62,11 @@ checkPayload spv e =
 
 modifyPayloadBitmap :: (Word16 -> Word16) -> BS.ByteString -> BS.ByteString
 modifyPayloadBitmap f bs =
-    let Right ((header, bitmap), rest) = S.runGetState ((,) <$> S.getWord8 <*> S.getWord16be) bs 0
+    let res = S.runGetState ((,) <$> S.getWord8 <*> S.getWord16be) bs 0
+        ((header, bitmap), rest) = case res of
+            Right v -> v
+            -- This does not happen
```
Comment (abizjak) [inline haskell-tests/Types/PayloadSerializationSpec.hs:68]: That is only true if the input bytestring has at least 24 bytes. State that as a precondition I suppose.

**Review by abizjak — APPROVED** [review]
Looks good. But we should not merge this before the other PRs are ready.

**Review by EmilLa1 — APPROVED** [review]
Looks good 👍 
I think the outstanding comment should be addressed before merging.

File: haskell-tests/Types/PayloadSerializationSpec.hs [hunk 2]
```diff
@@ -60,9 +60,15 @@ checkPayload spv e =
...
 
+-- Modify the bitmap portion of a payload. Assumes that the input bytestring
+-- is at least 24 bits long.
 modifyPayloadBitmap :: (Word16 -> Word16) -> BS.ByteString -> BS.ByteString
 modifyPayloadBitmap f bs =
-    let Right ((header, bitmap), rest) = S.runGetState ((,) <$> S.getWord8 <*> S.getWord16be) bs 0
+    let res = S.runGetState ((,) <$> S.getWord8 <*> S.getWord16be) bs 0
+        ((header, bitmap), rest) = case res of
+            Right v -> v
+            -- This happens only when bs is 23 bits or less.
```
Comment (abizjak) [inline haskell-tests/Types/PayloadSerializationSpec.hs:70]:
```suggestion
            -- This happens only when bs is 23 bytes or less.
```

Comment (vikt0r0) [reply haskell-tests/Types/PayloadSerializationSpec.hs:70]: No, 3 bytes right?

Comment (abizjak) [reply haskell-tests/Types/PayloadSerializationSpec.hs:70]: Yes, 3 bytes.  Better to use bytes than bits since that is what we are reading.

**Pull request merged**